docs/how-packages-and-artifacts-are-built.md

```diff
@@ -0,0 +1,114 @@
+* `buildinfo.json`: This file describes the code sources, the dependent packages, and the docker image in which the package will be built. This file can also declare a package as a service requiring state or a user account.
+* `build`: this is the script that will be executed inside a container of the image described in the buildinfo.json. There are a number of magic environment variables and paths (see below) that allow accessing the sources declared in the `buildinfo.json`, referencing configuration options and other packages, and exporting artifacts generated by the `build` script.
+
+Optionally, packages may have these folders:
+* `docker`: for customizing the docker image that the package is built in (see below)
+* `extra`: a folder of arbitrary data that will be volume mounted to the build container
+
+To actually get the artifacts generated in this process, they must be moved into the folder referenced by `$PKG_PATH` in the build container. The contents write-able, volume-mounted directory will be compressed into a package. This package can have folders called `bin` and `lib` that will be looped into the run-time `PATH` and `LD_LIBRARY_PATH`
+
+#### Sources
+Sources are singular artifacts or git repositories that are required to complete the package build. They are described via JSON maps with variable fields; the only guaranteed field is `kind`. There are a few kinds of sources:
+* `git`: Points to a git repository which can be private if the build host git client is configured correctly. Required fields: `kind`, `url`, `ref`, `ref_origin`. `ref` is the commit from the repo that should be built and `ref_origin` is the branch on which that commit live.
+* `git_local`: A filesystem-local git repository that may have been pulled in by a process outside gen/release. Required fields: `kind` and `rel_path`
+* `url`: URL of a singular artifact to download and volume mount inside the build container. Required fields: `kind`, `url`, `sha1`. `sha1` is the SHA-1 of the artifact and can be found by downloading the artifact and running `sha1sum <<my_artifact>>`
+* `url_extract`: URL is to an artifact that is a compressed file that will be automatically decompressed before volume mounting. Supported compression types are: `.tar.gz`, `.tgz`, and `.zip`. Required fields are: `kind`, `url`, and `sha1`
+
+As for describing in relation to the `buildinfo.json`:
+* If there are no required dependencies outside of the package directory, then no source information is required `buildinfo.json`
+* If there is only one source, a `source` field must be included in the `buildinfo.json`. The data from this source will be mounted inside the container as `/pkg/src/$PKG_NAME`
+* If there are multiple sources, a `sources` field must be provided in `buildinfo.json` and each of its subfields will be a source like the single source entry. Each of these sources will be mounted in the container as `/pkg/src/<<key name of source in JSON sources>>`
```

Comment:
:+1: To this being explained, remember it taking some time to figure it out the first time I was fighting with a buildinfo.json